### PR TITLE
Add solution builder script

### PR DESCRIPTION
**Pull request by hannahflroiter**
Added new `promote_sb_solution.py` script which can:
- Compile solution builder solution into `.ibsolution` file
- Upload the source `ibsolution` file to the target environment
- Deploy the source solution
- Deploy the target solution
- Upload and publish dependencies in the target environment (dependencies are now defined in an environment variable since we're not manually creating a `package.json` file)
- Set environment variables for github actions or azure devops pipelines
- Download the `ibsolution` file to local

Moved CI/CD template files into a new directory and updated them to install the library rather than having to copy code

Updated README with new sections for filesystem projects/solution builder projects as different environment variables are needed. Local flow (uploading code from a separate repository) is not supported for solution builder projects and it's assumed that deployed solutions is being used to publish the solution over marketplace.

### CODE REVIEW

File: ib_cicd/ib_helpers.py [hunk 1]
```diff
@@ -449,6 +500,9 @@ def deploy_solution(ib_host, api_token, ibsolution_path):
     headers = {"Authorization": "Bearer {0}".format(api_token)}
     url = f"{file_api_root}/solutions/deployed"
 
+    print("**!!!!!!!")
+    print(ibsolution_path)
```
Comment (BNJHope) [inline ib_cicd/ib_helpers.py:504]:
```suggestion
```

File: README.md [hunk 2]
```diff
@@ -62,9 +86,25 @@ If you're not using CI tooling and running the script locally you will need to s
     - **SOURCE_COMPILED_SOLUTIONS_PATH** Path to where `.ibsolution` files are generated. This will usually be one level higher than the `SOURCE_SOLUTION_DIR`
     - **TARGET_IB_HOST** e.g. "https://instabase.com"
     - **TARGET_IB_PATH** Path in the target IB environment where the solution will get uploaded
-3. Copy `cicd` folder into repository and move one of the `.yml` workflow files into a `.GitHub/workflows` folder
+3. Copy `cicd` folder into repository and move one of the `.yml` template workflow files into a `.GitHub/workflows` folder
```
Comment (BNJHope) [inline README.md:89]:
```suggestion
3. Copy `ib_cicd` folder into repository and move one of the `.yml` template workflow files into a `.GitHub/workflows` folder
```

File: README.md [hunk 3]
```diff
@@ -62,9 +86,25 @@ If you're not using CI tooling and running the script locally you will need to s
...
+2. Set git variables in the same place:
+    - **SOURCE_IB_HOST** e.g. "https://solution-eng.aws.sandbox.instabase.com"
+    - **SOURCE_WORKING_DIR** Path in source IB environment's filesystem to use for downloading dependencies and storing ibsolution files
+    - **TARGET_IB_HOST** e.g. "https://instabase.com"
+    - **TARGET_IB_PATH** Path in the target IB environment where the solution will get uploaded
+    - **SOLUTION_BUILDER_NAME** Name of solution builder project
+    - **FLOW_NAME** Name of flow to promote
+    - **WORKSPACE_DRIVE_PATH** Path in source IB environment's filesystem to the drive where the solution builder project is stored, e.g. hannahroiter/ci-cd/fs/Instabase Drive
+    - **DEPENDENCIES** String representation of list of model and marketplace package dependencies in the format `model_name==0.0.1,package_name==0.4.5,second_package==0.0.1`
+3. Copy `cicd` folder into repository and move one of the `.yml` template workflow files into a `.GitHub/workflows` folder
```
Comment (BNJHope) [inline README.md:105]:
```suggestion
3. Copy `ib_cicd` folder into repository and move one of the `.yml` template workflow files into a `.GitHub/workflows` folder
```

File: ib_cicd/migration_helpers.py [hunk 4]
```diff
@@ -391,3 +392,18 @@ def download_dependencies_from_dev_and_upload_to_prod(
...
+    Publish list of dependencies to marketplace
+    :param uploaded_ibsolutions: (list) list of paths to ibsolution dependency paths on the IB filesystem
+    :param ib_host: (string) IB host url (e.g. https://www.instabase.com)
+    :param api_token: (string) api token for IB environment
+    :return: None
+    """
+    for ib_solution_path in uploaded_ibsolutions:
+        publish_resp = publish_to_marketplace(ib_host, api_token, ib_solution_path)
+        logging.info(
+            "Publish response for {}: {}".format(ib_solution_path, publish_resp)
```
Comment (BNJHope) [inline ib_cicd/migration_helpers.py:408]:
```suggestion
            "Publish response for %s: %s", ib_solution_path, publish_resp
```

**Review by BNJHope — COMMENTED** [review]
Couple of comments, largely looks good 🙌 mentioned offline around paginating the list_directory API calls

File: ib_cicd/promote_sb_solution.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    paths = list_directory(ib_host, flows_path, ib_token)
+    for path in paths:
+        read_response = read_file_through_api(
+            ib_host, ib_token, os.path.join(path, "metadata.json")
+        ).content
+        metadata = json.loads(read_response)
+        if metadata["name"] == flow_name:
+            version = list_directory(ib_host, os.path.join(path, "versions"), ib_token)[
+                0
+            ]
```
Comment (gananPrabaharan) [inline ib_cicd/promote_sb_solution.py:88]:
```suggestion
            version = list_directory(ib_host, os.path.join(path, "versions"), ib_token)[0]
```
Is there a reason we take the first item from the list_directory output (is there always only one)?

File: ib_cicd/promote_sb_solution.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    :return: (string) IB filesystem path to flow version
+    """
+    flows_path = os.path.join(
+        ib_root, ".instabase_projects", solution_builder_name, "latest", "flows"
+    )
+    paths = list_directory(ib_host, flows_path, ib_token)
+    for path in paths:
+        read_response = read_file_through_api(
+            ib_host, ib_token, os.path.join(path, "metadata.json")
+        ).content
```
Comment (gananPrabaharan) [inline ib_cicd/promote_sb_solution.py:83]: I think at some point there was a change to the Solution Builder folder structure, we should probably figure out what versions of IB we support with the Solution Builder stuff. For example, I have this: 
https://platform.instabase.com/ganan.prabaharan/M_and_t/fs/Instabase%20Drive/.instabase_projects/MnTMortgage/latest/flows/MnT/, and it seems like the folder structure is different - there's no metadata.json

File: ib_cicd/promote_sb_solution.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+        for m in dependencies.split(",")
+    }
+
+
+def upload_icon(ib_host, ib_token, path, icon_path="icon.png"):
+    """
+    Uploads a local png file to the IB filesystem
+    :param ib_host: (string) IB host URL (e.g. https://platform.instabase.com/)
+    :param ib_token: (string) API token for IB environment
+    :param path: (string) IB path to upload icon
```
Comment (gananPrabaharan) [inline ib_cicd/promote_sb_solution.py:109]: Can we change this parameter to upload_path to avoid any confusion?

**Review by gananPrabaharan — COMMENTED** [review]
I went over it at a high level but didn't run the tests, I will try to schedule some time to do that tomorrow, but looks good so far. I think it might help us down the line if we make a list of the assumptions (folder structure, etc.), so that if anything changes, or if customers run into issues due to old IB versions it'd be a little easier for us to debug

File: ib_cicd/promote_sb_solution.py (hunk 6 above)
Comment (hannahflroiter) [reply ib_cicd/promote_sb_solution.py:83]: Difficult to figure out when the filesystem changed but I have the latest version working on v23.07 so added that as a note in the readme

**Review by gananPrabaharan — APPROVED** [review]

**Pull request merged**